component/http/middleware.go

```diff
@@ -123,6 +125,60 @@ func NewLoggingTracingMiddleware(path string) MiddlewareFunc {
 	}
 }
 
+type gzipResponseWriter struct {
+	io.Writer
+	http.ResponseWriter
+}
+
+// GZIPConfiguration defines configuration for GZIP Middleware.
+type GZIPConfiguration struct {
```

Comment:
Thanks for opening your first PR; nice work here! I really like the idea of excluding routes explicitly and using the header to decide the route behavior.

I took a first look and want to write some early comments; I'll come back with a more detailed review tomorrow. The more senior curators will pitch in as well, but here are my 2 cents.

- There has been an ongoing effort to keep the exported API surface as small as possible. 

Do you think we could un-export the *GZIPConfiguration* struct and its method? The builder pattern has used extensively throughout the project; a simple constructor and a method like `WithIgnoreRoutes(r ...string)` might be just as readable

```go
    gzipMiddleware := middleware.NewGzipMiddleware().WithIgnoreRoutes("/foo", "/bar").Build()
```

- Do you think there's a way to reduce the footgun potential, and is it worth doing so?

If someone defines the ignored routes with a trailing slash (eg. `/metrics/`, it will match *localhost:8080/metrics/* but not *localhost:8080/metrics?seconds=30*, which might be confusing.